File: src/Syncers/RetrieveRecords.cpp
#include "Syncers/RetrieveRecords.h"

#include <iostream>
#include <fstream>

#include <cpr/cpr.h>
#include <nlohmann/json.hpp>

#include "Utilities/Conversions.h"

using std::cout;
using std::endl;
using std::ofstream;

using Managers::CommitManager;
using Models::API;
using Models::Token;
using Utilities::Conversions;

namespace Syncers
{

#pragma region Constructors
RetrieveRecords::RetrieveRecords() { }
RetrieveRecords::RetrieveRecords(API api, Token token) 
    : token(token), api(api) { }
#pragma endregion

#pragma region Functions
void RetrieveRecords::retrieve(CommitManager::RetrieveTypes type)
{
    switch (type)
    {
        case CommitManager::RetrieveTypes::songs:
            fetchSongs();
            break;
        default:
            break;
    }
}
void RetrieveRecords::fetchSongs()
{
    cout<<"fetching songs"<<endl;
    auto url = api.url + "api/" + api.version + "/" + "song";

    auto auth = token.tokenType;
    auth.append(" " + token.accessToken);
    auto r = cpr::Get(cpr::Url{url},
         cpr::Header{{"authorization", auth},
             });

    if (r.status_code != (int)Result::OK) {
        cout<<"something went wrong\n";
        cout<<"status code: "<<r.status_code<<endl;
        cout<<"message: "<<r.text<<endl;
        return;
    }
    auto songData = nlohmann::json::parse(r.text);

    ofstream writeData{};
    writeData.open("songs.json");
    writeData<<songData.dump(4);
    writeData.close();
}
#pragma endregion

}

File: src/syncers/retrieve_records.rs
use std::default::Default;
use std::io::Error;

use crate::models;

#[derive(Default)]
pub struct RetrieveRecords {
    pub api: models::api::Api,
}

mod response {
    pub mod get_all_songs {
        #[derive(Debug, serde::Deserialize)]
        pub struct Response {
            pub message: String,
            pub data: Vec<icarus_models::song::Song>,
        }
    }
}

impl RetrieveRecords {
    pub async fn get_all_songs(
        &mut self,
        token: &icarus_models::token::AccessToken,
    ) -> Result<Vec<icarus_models::song::Song>, Error> {
        self.api.endpoint = String::from("api/v2/song/all");
        let url = format!("{}{}", self.api.url, self.api.endpoint);
        let access_token = token.bearer_token();

        println!("url: {url:?}");

        let client = reqwest::Client::builder().build().unwrap();
        let response = client
            .get(&url)
            .header(reqwest::header::AUTHORIZATION, access_token)
            .send()
            .await
            .unwrap();

        match response.status() {
            reqwest::StatusCode::OK => {
                // on success, parse our JSON to an API Response
                match response.json::<response::get_all_songs::Response>().await {
                    Ok(parsed) => {
                        println!("Response message: {:?}", parsed.message);
                        Ok(parsed.data)
                    }
                    Err(err) => Err(std::io::Error::other(err.to_string())),
                }
            }
            reqwest::StatusCode::UNAUTHORIZED => {
                Err(std::io::Error::other("Need to grab a new token"))
            }
            other => Err(std::io::Error::other(other.to_string())),
        }
    }
}
